src/lib.rs

```diff
@@ -124,96 +24,1138 @@ impl<T> RpcMessage for T where
-
-impl<T: transport::Listener<In = S::Req, Out = S::Res>, S: Service> Listener<S> for T {}
-
-#[cfg(feature = "flume-transport")]
-#[cfg_attr(quicrpc_docsrs, doc(cfg(feature = "flume-transport")))]
-/// Create a pair of [`RpcServer`] and [`RpcClient`] for the given [`Service`] type using a flume channel
-pub fn flume_channel<S: Service>(
-    size: usize,
-) -> (
-    RpcServer<S, server::FlumeListener<S>>,
-    RpcClient<S, client::FlumeConnector<S>>,
-) {
-    let (listener, connector) = transport::flume::channel(size);
-    (RpcServer::new(listener), RpcClient::new(connector))
-}
-
-#[cfg(feature = "test-utils")]
-#[cfg_attr(quicrpc_docsrs, doc(cfg(feature = "test-utils")))]
-/// Create a pair of [`RpcServer`] and [`RpcClient`] for the given [`Service`] type using a quinn channel
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+#[derive(Debug)]
+pub struct Client<M, R, S>(ClientInner<M>, PhantomData<(R, S)>);
+
+impl<M, R, S> Clone for Client<M, R, S> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone(), PhantomData)
+    }
+}
+
+impl<M, R, S> From<LocalSender<M, S>> for Client<M, R, S> {
+    fn from(tx: LocalSender<M, S>) -> Self {
+        Self(ClientInner::Local(tx.0), PhantomData)
+    }
+}
+
+impl<M, R, S> From<tokio::sync::mpsc::Sender<M>> for Client<M, R, S> {
+    fn from(tx: tokio::sync::mpsc::Sender<M>) -> Self {
+        LocalSender::from(tx).into()
+    }
+}
+
+impl<M, R, S> Client<M, R, S> {
+    #[cfg(feature = "rpc")]
+    pub fn quinn(endpoint: quinn::Endpoint, addr: std::net::SocketAddr) -> Self {
+        Self::boxed(rpc::QuinnRemoteConnection::new(endpoint, addr))
+    }
+
+    #[cfg(feature = "rpc")]
+    pub fn boxed(remote: impl rpc::RemoteConnection) -> Self {
+        Self(ClientInner::Remote(Box::new(remote)), PhantomData)
+    }
+
+    /// Get the local sender. This is useful if you don't care about remote
+    /// requests.
+    pub fn local(&self) -> Option<LocalSender<M, S>> {
+        match &self.0 {
+            ClientInner::Local(tx) => Some(tx.clone().into()),
+            ClientInner::Remote(..) => None,
+        }
+    }
+
+    /// Create a sender that allows sending messages to the service.
+    ///
+    /// In the local case, this is just a clone which has almost zero overhead.
+    /// Creating a local sender can not fail.
+    ///
+    /// In the remote case, this involves lazily creating a connection to the
+    /// remote side and then creating a new stream on the underlying
+    /// [`quinn`] or iroh connection.
+    ///
+    /// In both cases, the returned sender is fully self contained.
+    #[allow(clippy::type_complexity)]
+    pub fn request(
+        &self,
+    ) -> impl Future<
+        Output = Result<Request<LocalSender<M, S>, rpc::RemoteSender<R, S>>, RequestError>,
+    > + 'static
+    where
+        S: Service,
+        M: Send + Sync + 'static,
+        R: 'static,
+    {
+        #[cfg(feature = "rpc")]
```

Comment:
maybe place the cfg on the function level, given the whole body is different?

Reply:
I moved these things around a few times, but yes, you are probably right.

Reply:
Nah, on second thought, this fn has a very complex signature that I don't want to repeat.